### PR TITLE
Liquidation protocol fee

### PR DESCRIPTION
**Pull request by nope-finance**

### CODE REVIEW

File: token-lending/program/src/state/reserve.rs [hunk 1]
```diff
@@ -763,7 +785,9 @@ impl Pack for Reserve {
             config_deposit_limit,
             config_borrow_limit,
             config_fee_receiver,
+            config_protocol_liquidation_fee,
             _padding,
+            liquidity_accumulated_protocol_fees,
```
Comment (DaSichuan) [inline token-lending/program/src/state/reserve.rs:790]: Why are we doing it like this?

File: token-lending/program/src/state/reserve.rs [hunk 2]
```diff
@@ -290,10 +291,24 @@ impl Reserve {
...
+        let bonus_amount = Decimal::from(withdraw_amount).try_sub(base_amount)?;
+
+        // protocol takes a percentage of the bonus
+        let mut protocol_fee_amount = bonus_amount.try_mul(protocol_fee_rate)?.try_floor_u64()?;
+        // minimum protocol fee
+        if protocol_fee_amount == 0 {
+            protocol_fee_amount = 1;
+        }
+
+        let withdraw_amount = withdraw_amount.saturating_sub(protocol_fee_amount);
```
Comment (DaSichuan) [inline token-lending/program/src/state/reserve.rs:305]: Nit: why redefine `withdraw_amount` here?

File: token-lending/program/src/state/reserve.rs [hunk 3]
```diff
@@ -290,10 +291,24 @@ impl Reserve {
...
 
+        // calc the bonus amount
+        let base_amount = Decimal::from(withdraw_amount).try_div(bonus_rate)?;
+        let bonus_amount = Decimal::from(withdraw_amount).try_sub(base_amount)?;
+
+        // protocol takes a percentage of the bonus
+        let mut protocol_fee_amount = bonus_amount.try_mul(protocol_fee_rate)?.try_floor_u64()?;
+        // minimum protocol fee
+        if protocol_fee_amount == 0 {
+            protocol_fee_amount = 1;
```
Comment (DaSichuan) [inline token-lending/program/src/state/reserve.rs:302]: Doesn't enforcing a minimum protocol fee amount completely disincentive liquidators from liquidating sufficiently small liquidations?
I know they don't really go after the small ones anyway, but it seems like we might not want to have 0 incentivize liquidation opportunities maybe?
idk

Comment (DaSichuan) [reply token-lending/program/src/state/reserve.rs:302]: Also the liquidation will fail if withdraw_amount is 0

File: token-lending/program/src/processor.rs [hunk 4]
```diff
@@ -1557,6 +1561,7 @@ fn process_liquidate_obligation(
     let repay_reserve_liquidity_supply_info = next_account_info(account_info_iter)?;
     let withdraw_reserve_info = next_account_info(account_info_iter)?;
     let withdraw_reserve_collateral_supply_info = next_account_info(account_info_iter)?;
+    let withdraw_reserve_collateral_mint_info = next_account_info(account_info_iter)?;
```
Comment (DaSichuan) [inline token-lending/program/src/processor.rs:1564]: You didn't update the comment in instructions.rs that lists the input accounts

File: token-lending/program/src/instruction.rs [hunk 5]
```diff
@@ -1135,8 +1160,9 @@ pub fn liquidate_obligation(
             AccountMeta::new(destination_collateral_pubkey, false),
             AccountMeta::new(repay_reserve_pubkey, false),
             AccountMeta::new(repay_reserve_liquidity_supply_pubkey, false),
-            AccountMeta::new_readonly(withdraw_reserve_pubkey, false),
+            AccountMeta::new(withdraw_reserve_pubkey, false),
             AccountMeta::new(withdraw_reserve_collateral_supply_pubkey, false),
+            AccountMeta::new(withdraw_reserve_collateral_mint_pubkey, false),
```
Comment (DaSichuan) [inline token-lending/program/src/instruction.rs:1166]: Isn't this readonly?

File: token-lending/program/src/state/reserve.rs (hunk 1 above)
Comment (nope-finance) [reply token-lending/program/src/state/reserve.rs:790]: rather than?

File: token-lending/program/src/instruction.rs (hunk 5 above)
Comment (nope-finance) [reply token-lending/program/src/instruction.rs:1166]: true

File: token-lending/program/src/state/reserve.rs (hunk 3 above)
Comment (nope-finance) [reply token-lending/program/src/state/reserve.rs:302]: this only affects the smallest of liquidations, which no one is incentivized to do anyway.  wont liquidations already fail if withdraw amount is 0? (if you look in processor)

File: token-lending/program/src/instruction.rs (hunk 5 above)
Comment (nope-finance) [reply token-lending/program/src/instruction.rs:1166]: actually false LOL since there is a burn

**Comment by nope-finance** [comment]
#75 addresses this

**Pull request closed without merging**